### PR TITLE
Bach[Athena]: Support string functions for Athena

### PR DESCRIPTION
**Pull request by KathiaBarahona**
- Missing `get_dummies` TBD. Problem with column naming, since Athena does not support upper case in column names.
- ~Missing `to_json_array`~

Refactored `str.__get_item__` , since `athena` does not support `left` and `right` functions.

### CODE REVIEW

File: bach/tests/functional/bach/test_series_string.py [hunk 1]
```diff
@@ -33,14 +34,15 @@ def test_from_value(engine):
...
+@pytest.mark.athena_supported()
 def test_string_slice(engine):
     bt = get_df_with_test_data(engine)[['city']]
 
     slices = [
         # single values, keep small because we don't want to go out of range
         0, 1, 3, -3, -1,
         # some single value slices
-        slice(0), slice(1), slice(5), slice(-5), slice(-1),
+        slice(0, None), slice(1, None), slice(5, None), slice(-5, None), slice(-1, None),
```
Comment (KathiaBarahona) [inline bach/tests/functional/bach/test_series_string.py:45]: Seems this was not testing what it intended (previously `start` attr from the slice was None and `stop` attr actually had the value)

File: bach/tests/functional/bach/test_series_string.py [hunk 2]
```diff
@@ -58,29 +60,39 @@ def test_string_slice(engine):
...
-            bts2 = bt['city'].str.slice(s.start, s.stop)
+        if isinstance(s, slice):
+            _slice_1 = slice(s.start, s.stop)
+            _slice_2 = slice(s.start, s.stop)
         else:
-            bts1 = bt['city'].str[s]
-            bts2 = bt['city'].str.slice(s, s+1)
+            # str[-1] != str[-1:0]
+            _slice_1 = s
+            _slice_2 = slice(s, s + 1)
```
Comment (KathiaBarahona) [inline bach/tests/functional/bach/test_series_string.py:69]: there was a bug here, seems `str[s]` was not working in a pythonic way...

File: bach/bach/series/series_string.py [hunk 3]
```diff
@@ -24,64 +24,99 @@ class StringOperation:
...
+        if start_off > 0:
+            start_offset_expr = ConstValueExpression.construct(str(start_off))
+        else:
+            start_offset_expr = Expression.construct(
+                f'({{}} - {abs(start_off)})', self._base.str.len()
+            )
+
+        # Case 1: no substr length, therefore return all characters after start offset
+        # when stop == -1, it means the stop_offset is 0, therefore substr_length == 0
+        if stop is None or stop == -1:
```
Comment (KathiaBarahona) [inline bach/bach/series/series_string.py:66]: need to actually test `-1`

File: bach/tests/functional/bach/test_series_string.py (hunk 1 above)
Comment (KathiaBarahona) [reply bach/tests/functional/bach/test_series_string.py:45]: :thinking: will add both cases

File: bach/bach/series/series_string.py (hunk 3 above)
Comment (KathiaBarahona) [reply bach/bach/series/series_string.py]: fixed here https://github.com/objectiv/objectiv-analytics/pull/1189/commits/e8b2767249a37d7b0e42abb77f0b9bba598c626c

File: bach/tests/functional/bach/test_series_string.py (hunk 1 above)
Comment (KathiaBarahona) [reply bach/tests/functional/bach/test_series_string.py:46]: Added here and also with weird big indexes https://github.com/objectiv/objectiv-analytics/pull/1189/commits/e8b2767249a37d7b0e42abb77f0b9bba598c626c

File: bach/bach/series/series_string.py [hunk 4]
```diff
@@ -24,64 +24,129 @@ class StringOperation:
...
+
+        if isinstance(key, slice):
+            stop = key.stop
+        else:
+            # if key == -1, then we should just return last character, otherwise stop on the next character
+            stop = key + 1 if key != -1 else None
+
+        return self._base.copy_override(expression=self._get_substr_expression(start=start, stop=stop))
+
+    def _get_substr_expression(self, start=None, stop=None) -> Expression:
```
Comment (thijs-obj) [inline bach/bach/series/series_string.py:47]: Very much appreciate the comments, but still quite hard to review this.

I understand that different databases have slightly different implementations and that leads to some complications. But it's kinda hard to see whether all cases here actually cover all possible inputs.

**Question**: Have you considered not having the separate cases for 3, 4, and 5 and just a sql statement that works for all cases? something like below. This seems much simpler to me, and I don't quite see why it wouldn't work
```python
start_offset_expr = Expression.construct('greatest({}, 0)', start_offset_expr)  # make sure we don't get negative numbers
stop_offset_expr = # something similar to start_offset_expr here
substr_len_expr = Expression.construct('greatest({} - {}, 0)', stop_offset_expr, start_offset_expr)
return Expression.construct('substr({}, {}, {})', base_expr, start_offset_expr, substr_len_expr)
```

**Review by thijs-obj — CHANGES_REQUESTED** [review]
Nice improvement to the string slicing tests, but I think there are still some interesting cases missing. 

I am a big fan of porting the code to use `substr` so we have the same on all database platforms.

File: bach/bach/series/series_string.py (hunk 4 above)
Comment (KathiaBarahona) [reply bach/bach/series/series_string.py:46]: `greatest` helped a lot! took me some time, but managed to simplify it a lot. Done in https://github.com/objectiv/objectiv-analytics/pull/1189/commits/1909df24372dfb475d235d65288de43596c01ace

File: bach/tests/functional/bach/test_series_string.py [hunk 5]
```diff
@@ -41,14 +43,17 @@ def test_string_slice(engine):
...
+        slice(0, None), slice(1, None), slice(5, None), slice(-5, None), slice(-1, None),
         # simple slices
         slice(0, 3), slice(1, 3), slice(3, 3), slice(4, 3),
         # Some negatives
         slice(-4, -2), slice(-2, -2), slice(-2, 1), slice(1, -2),
         # some longer than some of the input strings
         slice(1, -8), slice(8, 1), slice(8, -4),
         # Some more with no beginnings or endings
-        slice(None, 3), slice(3, None), slice(None, -3), slice(-3, None)
+        slice(None, 3), slice(3, None), slice(None, -3), slice(-3, None),
```
[the remark this answers is not in the document]
Comment (KathiaBarahona) [reply bach/tests/functional/bach/test_series_string.py:54]: added https://github.com/objectiv/objectiv-analytics/pull/1189/commits/1909df24372dfb475d235d65288de43596c01ace

File: bach/bach/series/series_string.py [hunk 6]
```diff
@@ -156,6 +221,17 @@ def lower(self) -> 'SeriesString':
             expression=Expression.construct('lower({})', self._base)
         )
 
+    def len(self) -> 'SeriesInt64':
```
[the remark this answers is not in the document]
Comment (KathiaBarahona) [reply bach/bach/series/series_string.py:181]: done here https://github.com/objectiv/objectiv-analytics/pull/1189/commits/1909df24372dfb475d235d65288de43596c01ace

File: bach/bach/series/series_string.py [hunk 7]
```diff
@@ -24,64 +24,129 @@ class StringOperation:
...
-        :param item: an int for a single character, or a slice for some nice slicing
+        :param key: an int for a single character, or a slice for some nice slicing
         :return: SeriesString with the slice applied
         """
-        if isinstance(start, (int, type(None))):
-            item = slice(start, start + 1)
-        elif isinstance(start, slice):
-            item = start
+        if key and not isinstance(key, (int, slice)):
+            raise ValueError(f'Type not supported {type(key)}')
```
[the remark this answers is not in the document]
Comment (KathiaBarahona) [reply bach/bach/series/series_string.py]: https://github.com/objectiv/objectiv-analytics/pull/1189/commits/1909df24372dfb475d235d65288de43596c01ace

File: bach/bach/series/series_string.py (hunk 4 above)
Comment (thijs-obj) [reply bach/bach/series/series_string.py:46]: cool! much easier to read now :+1:

**Review by thijs-obj — APPROVED** [review]
lgtm

**Pull request merged**